Runner.cs

```diff
@@ -1,106 +1,168 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
 using System.IO;
 using System.Linq;
 using System.Net.Http;
 using System.Reflection;
 using System.Text;
+using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
+using MonoTouch.Dialog.Utilities;
 
 #nullable enable
 
 namespace SourceGeneratorPlayground
 {
-    internal class Runner
+    internal class Runner : IRunner
     {
-        private static List<MetadataReference>? s_references;
+        private static readonly Dictionary<string, MetadataReference>? s_references = new();
 
-        private readonly string _code;
-        private readonly string _generator;
+        private static readonly LRUCache<string, ImmutableArray<ISourceGenerator>> s_sourceGeneratorCache = new(entryLimit: 10);
+
+        private readonly string _baseUri;
 
         public string ErrorText { get; private set; } = "";
-        public string GeneratorOutput { get; private set; } = "";
+        public string? GeneratorOutput { get; private set; } = "";
         public string ProgramOutput { get; private set; } = "";
 
-        public Runner(string code, string generator)
+        public Runner(NavigationManager navigationManager)
         {
-            _code = code;
-            _generator = generator;
+            _baseUri = navigationManager.BaseUri;
         }
 
-        internal async Task Run(string baseUri)
+        public async Task RunAsync(string code, string generator, CancellationToken cancellationToken)
         {
-            if (s_references == null)
-            {
-                s_references = await GetReferences(baseUri);
-            }
+            await UpdateReferences(_baseUri);
```

Comment:
This is one of the major differences compared to PR#9: on every run, we quickly check that all references are up-to-date. To make that less cumbersome, the `s_references` field is now a dictionary, not a list (key is the location).